### ISSUE TITLE
[css-overflow-5]: calling focus on the carousel / scroll container

### ISSUE THREAD

**Issue by smhigley** [w3c/csswg-drafts#12240]
## Current behavior

Adding `::scroll-marker-group` and scroll markers to an element appears to make it focusable but not in the tab order, similar to adding `tabindex="-1"` to the element.

Saving a reference to the scroll container and calling `element.focus()` moves focus to the scroll container itself, not to either the inner scroll buttons or scroll marker group. This can be verified either by running a screen reader or inspecting the related FocusEvent's `event.target` (which currently points to pseudo-elements when you tab to them, but points to the scroll container if you programmatically call focus).

## Problem

There are two related issues around handling `.focus()` calls, one with the current behavior, and the other with the API in general.

### 1. Current behavior

It seems pretty clear to me that making an element into a scroll container does not imply an authorial intent for that element to itself be interactive or exposed to assistive tech users. Instead, the scroll buttons, markers, and scrolling content are the semantically relevant controls. In this basic case, I don't think the scroll container element should become focusable just because of `::scroll-marker-group` -- it likely does not have (and often _should_ not have) an interactive role and accessible name that would support it being focusable and accessible.

By default, it seems like the best course of action would be to either set focus to the first internal focusable element (which is currently a scroll button if it exists, and the scroll markers if it doesn't), or always set it to the scroll markers. I'm curious on other folks' thoughts on which it should be, since generally I would lean towards moving focus to the scroll markers, but having the scroll buttons before them in the tab order complicates that.

I personally think authors need the ability to call focus on the prev/next scroll buttons and scroll marker group separately, so perhaps figuring out how to expose that would solve this question (and the next) as well.

### 2. Resolving `.focus()` on focusable scroll containers

This problem extends beyond resolving the current issue with focusing what are otherwise unfocusable scroll containers. If an author has intentionally created a focusable scroll container, then calling `scrollContainerElement.focus()` would need to send focus directly to the element. Authors would still need a way to send focus to the scroll marker/button controls though, separately from the container element.

Since pseudo-elements are not currently accessible in JS, and also are type Element and do not support the `.focus()` method, it probably needs someone more creative than I am right now to figure out how this would be solved 😅

(As a brief top-of-mind example on why this would be necessary -- consider an overflow container with `tabindex="-1"` to handle fallback focus if all content is deleted and there is no longer overflow. However, when sending focus to the region while it is interactive, the author would want to move focus directly to the scroll buttons/markers. Or a scenario where the author has made custom keyboard commands to jump to a specific slide, and needs to manage focus when doing so.)

**Comment by smhigley** [comment]
@flackr (well I now feel awkward trying to write 6 different comments tagging you in with slightly different wording)

**Comment by flackr** [comment]
> It seems pretty clear to me that making an element into a scroll container does not imply an authorial intent for that element to itself be interactive or exposed to assistive tech users.

This isn't related to `scroll-marker-group`, but rather just making it a scroll container, see https://chromestatus.com/feature/5231964663578624. I.e. just having a scroll container makes it focusable https://jsbin.com/yufibuw/edit?html,css,js,output

@mfreed7 you're listed an owner on this feature and maybe have more context about where this was discussed.

**Comment by smhigley** [comment]
@flackr I'm curious if you think the `scroll-marker-group` behavior should also trigger focusability in the same way that regular scroll should. The reason for it happening with a traditional scroll overflow element is so keyboard users can scroll, but that doesn't really apply to scroll marker containers. Mouse scrolling doesn't do anything, and keyboard arrows don't scroll in the traditional way.

That said, the primary issue is still figuring out how authors can call focus directly on the scroll markers, separately from the container element.

**Comment by flackr** [comment]
> but that doesn't really apply to scroll marker containers. Mouse scrolling doesn't do anything, and keyboard arrows don't scroll in the traditional way.

But it is just a scrolling element.

I'm confused, adding scroll-marker-group does not affect the focusability of an element as far as I know.

**Comment by flackr** [comment]
> That said, the primary issue is still figuring out how authors can call focus directly on the scroll markers, separately from the container element.

@danielsakhapov (https://github.com/w3c/csswg-drafts/issues/12163) and @naomr (https://github.com/w3c/csswg-drafts/issues/12575) have both proposed ways that we might want to handle events on pseudo-elements. Either of these strategies could lend themselves to allowing setting focus, e.g.
* If we have a CSSPseudoElement, you can call focus on it.
* We could have a pseudoElement target string on the focus method.
* If there is a DOM element proxy for it you could call focus on this.

**Comment by mfreed7** [comment]
> This isn't related to `scroll-marker-group`, but rather just making it a scroll container, see https://chromestatus.com/feature/5231964663578624. I.e. just having a scroll container makes it focusable https://jsbin.com/yufibuw/edit?html,css,js,output
> 
> [@mfreed7](https://github.com/mfreed7) you're listed an owner on this feature and maybe have more context about where this was discussed.

Right - starting with Chrom(ium) 132, scroll containers are focusable, and they become keyboard focusable (part of the sequential focus navigation order) if they do not contain any other focusable descendants.



> [@flackr](https://github.com/flackr) I'm curious if you think the `scroll-marker-group` behavior should also trigger focusability in the same way that regular scroll should. The reason for it happening with a traditional scroll overflow element is so keyboard users can scroll, but that doesn't really apply to scroll marker containers. Mouse scrolling doesn't do anything, and keyboard arrows don't scroll in the traditional way.

This is an interesting point. Do you think we should modify the above heuristic in some way to avoid that treatment for scroll marker containers? That would (I think?) be relatively easy to do, if it's the correct thing to do.

**Comment by smhigley** [comment]
I think my initial confusion around it triggering focusability was caused by a personal issue with janky horizontal scroll behavior when I was testing, so I thought it just didn't support mouse/keyboard scroll 😅. Given it does, that seems OK.

The central issue is still the authoring need to be able to call focus on the container separately from the scroll markers, together with the following:
- When focus is on the scroll markers, it should be possible to find the current and next/previousfocusable element through using `document.activeElement`
- When looking for the next/previous focusable element starting from either something within the carousel or after it, it should be possible to query & find the scroll markers separately from the container (also based on where they are in the DOM in relation to other focusable elements, which is different from where the container is)

From the perspective of a library maintainer, those should ideally be possible for every possible way to author a carousel, since the team writing focus management code is often not the team writing the carousel code.

**Comment by smhigley** [comment]
I thought it might be useful to add a concrete example of where this might be needed:

Sometimes a web app has multiple discrete sections (e.g. an editing surface, a toolbar, and a sidepane) where each section has its own tab loop, with F6 (or ctrl + F6) moving focus between sections. When authoring both the tab loop and the F6 focus target, it would be important to be able to query focusables and set focus on the most relevant focus target depending on the forwards/backwards direction.

Similarly, there would be a need to manage tabindex / tabbability for both the scroll container and the scroll markers separately.
